src/platform/storage/zephyr/nvs/storage.c

```diff
@@ -163,41 +211,85 @@ crc_check(const unsigned char *data, const size_t data_len) {
+           tests/debug builds should fail hard. */
+        assert(false);
+
+        mender_log_error("Failed to get storage spec");
+        return MENDER_FAIL;
+    }
+#else
+    /* With custom storage, we need to call mender_get_storage_spec() to get the
+       sector offset and only then we'll know how many offsets we need to work
+       with. */
+    if (MENDER_OK != mender_get_storage_spec(&(mender_storage_nvs_handle.flash_device), &part_id, &sector_offset)) {
+        mender_log_error("Failed to get custom storage spec");
+        return MENDER_FAIL;
+    }
+    n_sectors = sector_offset + CONFIG_MENDER_STORAGE_NVS_SECTOR_COUNT + CONFIG_MENDER_STORAGE_DEPLOYMENT_LOGS_SECTORS;
+    if (NULL == (flash_sectors = mender_calloc(sizeof(struct flash_sector), n_sectors))) {
+        mender_log_error("Cannot allocate memory");
+        return MENDER_FAIL;
+    }
+#endif /* CONFIG_MENDER_STORAGE_PARTITION_CUSTOM */
 
     /* Get flash info */
-    mender_storage_nvs_handle.flash_device = MENDER_STORAGE_DEVICE;
+    /** First check if the device is ready. */
     if (!device_is_ready(mender_storage_nvs_handle.flash_device)) {
         mender_log_error("Flash device not ready");
         return MENDER_FAIL;
     }
 
-    n_sectors = sizeof(flash_sectors) / sizeof(flash_sectors[0]);
-    result    = flash_area_get_sectors(MENDER_STORAGE_FLASH_AREA, &n_sectors, flash_sectors);
+    /** Then get the information about the partition offset on the device
+        (flash_sectors[0].fs_off below is an offset relative to the partition
+        start). */
+    if (0 != (result = flash_area_open(part_id, &fap))) {
+        mender_log_error("Unable to open the Mender storage flash area");
+        return MENDER_FAIL;
+    }
+    part_offset = fap->fa_off;
+    flash_area_close(fap);
+
+    /** Then get information about the individual sectors. */
+    const uint32_t n_sectors_required = n_sectors; /* optimized out for sure */
+    result                            = flash_area_get_sectors(part_id, &n_sectors, flash_sectors);
     if ((0 != result) && (-ENOMEM != result)) {
         /* -ENOMEM means there were more sectors than the supplied size of the
             sector array (flash_sectors), but we don't worry about that, we just
             need the info about the first N sectors we want to use. */
         mender_log_error("Failed to get info about flash sectors in the Mender flash area [%d]", -result);
         return MENDER_FAIL;
     }
-    if (n_sectors != (sizeof(flash_sectors) / sizeof(flash_sectors[0]))) {
-        mender_log_error(
-            "Not enough sectors on the flash for Mender (required: %" PRIu32 "d, available: %zu)", sizeof(flash_sectors) / sizeof(flash_sectors[0]), n_sectors);
+    if (n_sectors != n_sectors_required) {
+        mender_log_error("Not enough sectors on the flash for Mender (required: %" PRIu32 "d, available: %" PRIu32 "d)", n_sectors_required, n_sectors);
```

Comment:
Pre-existing, but the trailing `d`s here are wrong (right?). The `PRIu32` is the correct and complete format qualifier. I guess it is just printing a literal `d` at the end.

Reply:
Right! The `intN_t` man page says these are *length modifiers* which, to me, says they should be followed by a *conversion specifier* as the `printf(3)` man page says, but that's not the case. Good catch! And a reminder I should test this case manually.

Reply:
(update: works fine with the `d`s dropped )